grml-debootstrap

```diff
@@ -1581,74 +1580,102 @@ grub_install() {
     einfo "Installing Grub as bootloader."
 
-    if ! clean_chroot "${MNTPOINT}" dpkg --list grub-pc 2>/dev/null | grep -q '^ii' ; then
-      echo "Notice: grub-pc package not present yet, installing it therefore."
-      # shellcheck disable=SC2086
-      clean_chroot "$MNTPOINT" DEBIAN_FRONTEND=$DEBIAN_FRONTEND apt-get -y --no-install-recommends install $DPKG_OPTIONS grub-pc
+    local grub_pc_package_name=''
+
+    if [ -z "$VMEFI" ]; then
+      grub_pc_package_name='grub-pc'
+    elif [ "$ARCH" != 'arm64' ]; then
+      # We install grub-pc-bin instead of grub-pc when EFI is enabled, because
+      # otherwise the EFI bootloader won't be automatically updated when GRUB
+      # packages are uploaded. Doing this means that the BIOS bootloader won't
+      # be automatically updated, which stinks, however the BIOS bootloader
+      # doesn't have the same security concerns as the EFI bootloader (there's
+      # no Secure Boot to grapple with when using legacy BIOS boot) so it's
+      # better to let the BIOS bootloader lag behind and update the EFI one
+      # than to let the EFI bootloader lag behind and update the BIOS one.
+      grub_pc_package_name='grub-pc-bin'
+    fi
+
+    if [ -n "${grub_pc_package_name}" ]; then
+      if ! clean_chroot "${MNTPOINT}" dpkg --list "${grub_pc_package_name}" 2>/dev/null | grep -q '^ii' ; then
+        echo "Notice: ${grub_pc_package_name} package not present yet, installing it therefore."
+        clean_chroot "$MNTPOINT" DEBIAN_FRONTEND=$DEBIAN_FRONTEND apt-get -y --no-install-recommends install $DPKG_OPTIONS "${grub_pc_package_name}"
+      fi
     fi
 
     mkdir -p "${MNTPOINT}/boot/grub"
-    if ! [ -d "${MNTPOINT}"/usr/lib/grub/i386-pc/ ] ; then
-      eerror "Error: grub not installed inside Virtual Machine. Can not install bootloader."
-      bailout 1
+    if [ -d "${MNTPOINT}"/usr/lib/grub/i386-pc/ ]; then
+      cp -a "${MNTPOINT}"/usr/lib/grub/i386-pc "${MNTPOINT}/boot/grub/"
+    elif [ "$ARCH" != 'arm64' ]; then
+       eerror "Error: grub not installed inside Virtual Machine. Can not install bootloader."
+       bailout 1
     fi
-    cp -a "${MNTPOINT}"/usr/lib/grub/i386-pc "${MNTPOINT}/boot/grub/"
 
     if [ -n "$VMEFI" ]; then
-
       mkdir -p "${MNTPOINT}"/boot/efi
       mount -t vfat "${EFI_TARGET}" "${MNTPOINT}"/boot/efi
 
+      if [ -z "${EFI_ID}" ]; then
+        EFI_ID="$(
+          source "${MNTPOINT}"/etc/default/grub
+          for file in "${MNTPOINT}"/etc/default/grub.d/*.cfg; do
+            if [ -f "${file}" ]; then
+              source "${file}"
+            fi
+          done
+          echo "${GRUB_DISTRIBUTOR}"
+        )"
+      else
+        # This is prefixed with a `z` so that it takes precedence over any other GRUB config.
+        echo "GRUB_DISTRIBUTOR='${EFI_ID}'" > "${MNTPOINT}"/etc/default/grub.d/z-grml-debootstrap-efi-id.cfg
+      fi
+
+      mkdir -p "${MNTPOINT}"/boot/efi/EFI/BOOT
+      mkdir -p "${MNTPOINT}"/boot/efi/EFI/"${EFI_ID}"
```

Comment:
can we just skip this?

Reply:
We cannot. If those directories do not exist, the grub-efi-ARCH package **will not install the bootloader at package installation time.** This is coded into the postinst script's logic. Creating these dirs allows the package to auto-install (and more importantly, auto-update) the bootloader.

Reply:
that will at least need a comment

Reply:
could you please point me to the code for this in the grub2 package? I'm not seeing it in https://sources.debian.org/src/grub2/2.12-5/debian/postinst.in/

Reply:
I'll try to find it again, I'm certain I saw it and even verified that the mkdir commands worked to fix the bug, but now I'm having trouble finding it too. It might be only a Bookworm thing.

Reply:
nvm, found it, it looks like this was fixed in >Bookworm but is a problem in Bookworm specifically. https://sources.debian.org/src/grub2/2.06-13%2Bdeb12u1/debian/postinst.in/#L720

Reply:
ok, i understand the code path in the postinst for updates.

but this will be done correctly by grub-install --target=x86_64-efi  / grub-install --target=arm64-efi - no?

don't get me wrong, i don't want to be annoying about this. I just really don't want to have code in grml-debootstrap dealing with internals of the grub packaging, and very much not code that the grub maintainers do not know about and do not include in their testing of upgrade paths.

Reply:
Erm, I thought you had explicitly asked me to remove the grub-install commands: https://github.com/grml/grml-debootstrap/pull/299#issuecomment-2585163230 This PR currently removes them in favor of making the Debian packaging do the right thing automatically. If I misunderstood you, I'm happy to undo that.

Reply:
Quote [grub2 package, Debian changelog, Colin Watson Nov 2020](https://metadata.ftp-master.debian.org/changelogs//main/g/grub2/grub2_2.12-5_changelog)

>   * Don't call grub-install on fresh install of grub-pc.  It's the job of
>    installers to do that after a fresh install.
>
> -- Colin Watson <cjwatson@debian.org>  Sun, 08 Nov 2020 16:26:08 +0000

Reply:
Quote [grub2 package, Debian changelog, Colin Watson Nov 2020](https://metadata.ftp-master.debian.org/changelogs//main/g/grub2/grub2_2.12-5_changelog)

>   * Don't call grub-install on fresh install of grub-pc.  It's the job of
>    installers to do that after a fresh install.
>
> -- Colin Watson <cjwatson@debian.org>  Sun, 08 Nov 2020 16:26:08 +0000